proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -293,70 +262,234 @@ protocol TargetBuildContext {
+    /// Every executable product has exactly one main executable target.
+    /// This target will always be one of the targets in the product's
+    /// `targets` array.
+    let mainTarget: Target
+}
+
+/// Represents a library product defined in a package.
+class LibraryProduct: Product {
+    /// Whether the library is static, dynamic, or whether the type is
+    /// determined automatically.
+    let type: LibraryType
+
+    /// A type of library product.
+    enum LibraryType {
+        /// Static library.
+        case `static`
+
+        /// Dynamic library.
+        case `dynamic`
+
+        /// The type of library is unspecified and will be determined
+        /// at build time.
+        case automatic
+    }
+}
+
+/// Represents a target defined in a package. Specializations represent
+/// different types of targets.
+class Target {
+    /// The name of the target, as defined in the package manifest. It is
+    /// unique among the targets of the package.
+    let name: String
+    
+    /// The absolute path of the target directory in the local file system.
+    let directory: Path
+    
+    /// Any other targets on which this target depends, in the same order
+    /// as specified in the package manifest. Any conditional dependencies
+    /// that do not apply will have already been filtered out.
+    let dependencies: [Dependency]
+
+    /// Represents a dependency of a target on a product or on another target.
+    enum Dependency {
+        /// A dependency on a target in the same package.
+        case target(Target)
+
+        /// A dependency on a product in another package.
+        case product(Product, package: Package)
     }
 }
 
-/// Constructs commands to run during the build, including command lines,
-/// environment variables, initial working directory, etc. All paths should
-/// be based on the ones passed to the plugin in the target build context.
-protocol CommandConstructor {
+/// Represents a target consisting of a source code module, containing either
+/// Swift or source files in one of the C-based languages.
+class SourceModuleTarget: Target {
+    /// The name of the module produced by the target (derived from the target
+    /// name, though future SwiftPM versions may allow this to be customized).
+    let moduleName: String
+    
+    /// The directory containing public C headers, if applicable. This will
+    /// only be set for targets that have Clang sources, and only if there is
+    /// a public headers directory.
+    let publicHeadersDirectory: Path?
+ 
+    /// The source files that are associated with this target (any files that
+    /// have been excluded in the manifest have already been filtered out).
+    let sourceFiles: FileList
+}
+
+/// Represents a target describing a system library that is expected to be
+/// present on the host system.
+class SystemLibraryTarget: Target {
+    /// The directory containing public C headers, if applicable.
+    let publicHeadersDirectory: Path?
```

Comment:
@neonichu What else should we expose to plugins about system library targets?  The idea here is to prepare for future types of plugins by exposing a distilled version of the package graph, so we won't need to change this when proposing other capabilities than build tools.  So what would make sense to expose here do you think?  The main reason for having the headers directory here is if build tool plugins need to add search paths to it, etc.

Reply:
I think we also need compiler and linker flags. If the library has an underlying pkgConfig file, we retrieve those settings from it and feed them into the normal build process. I could see that this might also be something a build tool plugin needs to do.

Reply:
Thanks — that makes sense.  It looks as if this only happens during a build, after plugins are invoked, and is not available in general when plugins run.  Would we need to start doing that earlier in order to support this?

Reply:
I think so, yes. Though I don't think exposing this is necessary critical for the initial release of plugins.

Reply:
Agreed.  I have pared back the changes in this amendment (I kind of forgot that it was just an amendment TBH, and got a bit carried away with proposed changes).  But this is really useful information for a second round of more advanced API in the future.  Thanks!